pennylane/_device.py

```diff
@@ -178,13 +178,45 @@ def shots(self, shots):
         Raises:
             DeviceError: if number of shots is less than 1
         """
-        if shots < 1:
+        if isinstance(shots, int):
+            if shots < 1:
+                raise DeviceError(
+                    "The specified number of shots needs to be at least 1. Got {}.".format(shots)
+                )
+
+            self._shots = int(shots)
+            self._shot_vector = None
+
+        elif isinstance(shots, Sequence):
+            # a sequence of shots has been provided
+
+            if all(isinstance(s, int) for s in shots):
+
+                if len(set(shots)) == 1:
+                    # All shots are identical; represent the shot vector
+                    # in a sparse format.
+                    shot_vector = [(shots[0], len(shots))]
+                    total_shots = shots[0] * len(shots)
+                else:
+                    # Iterate through the shots, and group consecutive identical shots
+                    split_at_repeated = np.split(shots, np.where(np.diff(shots) != 0)[0] + 1)
+                    shot_vector = [
+                        i.item() if len(i) == 1 else (i[0], len(i)) for i in split_at_repeated
+                    ]
+                    total_shots = np.sum(shots)
```

Comment:
Could outsource this procedure to a function, so it can be easily tested?

Reply:
Good idea, will add!